collator/src/lib.rs

```diff
@@ -215,197 +210,204 @@ where
 
-	fn produce_candidate(
+	fn build_collation(
 		&mut self,
-		relay_chain_parent: PHash,
-		global_validation: GlobalValidationData,
-		local_validation: LocalValidationData,
-		downward_messages: Vec<DownwardMessage>,
-	) -> Self::ProduceCandidate {
-		let factory = self.proposer_factory.clone();
-		let inherent_providers = self.inherent_data_providers.clone();
-		let block_import = self.block_import.clone();
-
-		trace!(target: "cumulus-collator", "Producing candidate");
-
-		let last_head = match HeadData::<Block>::decode(&mut &local_validation.parent_head.0[..]) {
-			Ok(x) => x,
+		block: ParachainBlockData<Block>,
+		block_hash: Block::Hash,
+	) -> Option<Collation> {
+		let block_data = BlockData(block.encode());
+		let header = block.into_header();
+		let head_data = HeadData(header.encode());
+
+		let state = match self.backend.state_at(BlockId::Hash(block_hash)) {
+			Ok(state) => state,
 			Err(e) => {
-				error!(target: "cumulus-collator", "Could not decode the head data: {:?}", e);
-				return Box::pin(future::ready(None));
+				error!(target: "cumulus-collator", "Failed to get state of the freshly build block: {:?}", e);
+				return None;
 			}
 		};
 
-		if !self.check_block_status(last_head.header.hash()) {
-			return future::ready(None).boxed();
-		}
+		state.inspect_state(|| {
+			let upward_messages = sp_io::storage::get(well_known_keys::UPWARD_MESSAGES);
+			let upward_messages = match upward_messages.map(|v| Vec::<UpwardMessage>::decode(&mut &v[..])) {
+				Some(Ok(msgs)) => msgs,
+				Some(Err(e)) => {
+					error!(target: "cumulus-collator", "Failed to decode upward messages from the build block: {:?}", e);
+					return None
+				},
+				None => Vec::new(),
+			};
 
-		let proposer_future = factory.lock().init(&last_head.header);
-
-		let wait_to_announce = self.wait_to_announce.clone();
-
-		Box::pin(async move {
-			let proposer = proposer_future
-				.await
-				.map_err(|e| {
-					error!(
-						target: "cumulus-collator",
-						"Could not create proposer: {:?}",
-						e,
-					)
-				})
-				.ok()?;
-
-			let inherent_data = Self::inherent_data(
-				inherent_providers,
-				global_validation,
-				local_validation,
-				downward_messages,
-			)?;
-
-			let Proposal {
-				block,
-				storage_changes,
-				proof,
-			} = proposer
-				.propose(
-					inherent_data,
-					Default::default(),
-					//TODO: Fix this.
-					Duration::from_millis(500),
-					RecordProof::Yes,
-				)
-				.await
-				.map_err(|e| {
-					error!(
-						target: "cumulus-collator",
-						"Proposing failed: {:?}",
-						e,
-					)
-				})
-				.ok()?;
-
-			let proof = match proof {
-				Some(proof) => proof,
-				None => {
-					error!(
-						target: "cumulus-collator",
-						"Proposer did not return the requested proof.",
-					);
+			let new_validation_code = sp_io::storage::get(well_known_keys::NEW_VALIDATION_CODE);
 
+			Some(Collation {
+				//TODO: What should we do here?
+				fees: 0,
```

Comment:
The fees member can just be removed. We don't have anything that fees are charged for anymore, now that we use channels for messages.

Reply:
This comes from the Polkadot side.

Reply:
Yeah, we can remove it there. It's plumbed everywhere but essentially it starts it's life here. 

https://github.com/paritytech/polkadot/blob/4e5df2988ce13851faf56de4e2f85b68a5601b8c/node/core/candidate-validation/src/lib.rs#L494-L494

so it's just a constant